src/treebuilders/apply.cpp

```diff
@@ -308,6 +461,46 @@ template <int D> void apply(FunctionTree<D> &out, DerivativeOperator<D> &oper, F
     print::separator(10, ' ');
 }
 
+template <int D> void apply(CompFunction<D> &out, DerivativeOperator<D> &oper, CompFunction<D> &inp, int dir, ComplexDouble metric[4][4]) {
+    //TODO: sums and not only each components independently
+    ComplexDouble defaultMetric[4][4];
+    for (int i=0; i<4; i++){
+        for (int j=0; j<4; j++){
+            if (i==j) defaultMetric[i][j] = 1.0;
+            else defaultMetric[i][j] = 0.0;
+        }
+    }
+    if (metric == nullptr) {
+        metric = defaultMetric;
+    }
+    for (int icomp = 0; icomp < inp.Ncomp(); icomp++){
+        for (int ocomp = 0; ocomp < 4; ocomp++){
+            if (std::norm(metric[icomp][ocomp]) > MachinePrec) {
+                if (inp.isreal() and (std::imag(metric[icomp][ocomp]) < MachinePrec or inp.Ncomp() == 1) ) {
+                    apply(*out.CompD[ocomp], oper, *inp.CompD[icomp], dir);
+                    if (std::norm(metric[icomp][ocomp] - 1.0) > MachinePrec) {
+                        if(std::imag(metric[icomp][ocomp]) < MachinePrec) out.CompD[ocomp]->rescale(std::real(metric[icomp][ocomp]));
+                        else out.func_ptr->data.c1[ocomp] *= metric[icomp][ocomp]; //TODO: multiply c1 in rescale?
```

Comment:
TODO still relevant?

Reply:
I rephrased: // To consider: multiply c1 in rescale?
 The "c1 constant" is somewhat new and it is not yet completely integrated into the logic. It is a "soft" multiplicative constant; soft meaning the same as the conjugate flag for a function: not explicitly applied before it is required. 
 So far it is used for the complex spin operators: we do not want to convert all real functions to complex just because there is a complex multiplicative factor in the operator.